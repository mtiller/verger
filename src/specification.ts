export interface ASTUnionType {
  type: "union";
  name: string;
  subtypes: string[];
}

export function isUnion(a: ASTTree): a is ASTUnionType {
  return a.type === "union";
}

export interface ASTBaseType {
  type: "base";
  name: string;
  extends: string[];
  fields: Map<string, string | string[]>;
}

export function isBase(a: ASTTree): a is ASTBaseType {
  return a.type === "base";
}

export interface ASTLeafType {
  type: "leaf";
  tag: string;
  name: string;
  extends: string[];
  fields: Map<string, string | string[]>;
}

export function isLeaf(a: ASTTree): a is ASTLeafType {
  return a.type === "leaf";
}

export type ASTTree = ASTUnionType | ASTBaseType | ASTLeafType;
export interface ASTSpec {
    tagName: string;
    names: Set<string>;
    unions: Map<string, ASTUnionType>;
    bases: Map<string, ASTBaseType>;
    leaves: Map<string, ASTLeafType>;
}

export function createASTTree(spec: any): ASTSpec {
  const types: ASTSpec = {
      tagName: "tag",
      names: new Set(),
      unions: new Map(),
      bases: new Map(),
      leaves: new Map()
  }
  if (spec.hasOwnProperty("nodes")) {
    const root = spec["nodes"];
    for (const [name, content] of Object.entries(root)) {
      processSpecNode(name, content, types);
    }
    return types;
  } else {
    throw new Error("Missing nodes field");
  }
}

function processSpecNode(
  n: string,
  content: any,
  types: ASTSpec,
): ASTTree {
  if (content === null || content === undefined)
    throw new Error(`Missing content for ${name}`);
  if (Array.isArray(content)) {
    const base = n.startsWith("^");
    const ret: ASTLeafType | ASTBaseType = base ? {
        type: "base",
        name: n.slice(1),
        extends: [],
        fields: new Map<string, string>()
    } : {
        type: "leaf",
        name: n,
        tag: n.toLowerCase(),
        extends: [],
        fields: new Map<string, string>()
    };
    for (const item of content) {
      const keys = Object.keys(item);
      if (keys.length !== 1) {
        throw new Error(
          `Unexpected field type specification: ${JSON.stringify(item)}`
        );
      } else {
        const fieldName = keys[0];
        const type = item[fieldName];
        if (typeof type === "string") {
          if (fieldName === "extends") {
            const base = types.bases.get(type);
            if (base === undefined) {
              throw new Error(
                `Type ${ret.name} cannot extend from unknown (not yet defined) base type ${type}`
              );
            }
            if (base.type !== "base") {
              throw new Error(
                `Node ${ret.name} cannot extend from non-base type ${type}`
              );
            }
            ret.extends.push(type);
          } else {
            ret.fields.set(fieldName, type);
          }
        } else if (Array.isArray(type)) {
          if (type.every((x) => typeof x === "string")) {
            ret.fields.set(fieldName, type as string[]);
          }
        } else {
          throw new Error(
            `Unexpected value for field ${fieldName}: ${JSON.stringify(type)}}`
          );
        }
      }
    }
    if (types.names.has(ret.name)) {
      throw new Error(`Multiple definitions for type ${ret.name}`);
    }
    types.names.add(ret.name);
    if (ret.type==="base") types.bases.set(ret.name, ret);
    else types.leaves.set(ret.name, ret);
    return ret;
  } else {
    const subtypes = Object.keys(content);
    const ret: ASTUnionType = {
      type: "union",
      name: n,
      subtypes,
    };
    for (const [subtype, contents] of Object.entries(content)) {
      ret.subtypes.push(subtype);
      processSpecNode(subtype, contents, types);
    }
    if (types.names.has(ret.name)) {
      throw new Error(`Multiple definitions for type ${name}`);
    }
    types.names.add(ret.name);
    types.unions.set(ret.name, ret);
    return ret;
  }
}
