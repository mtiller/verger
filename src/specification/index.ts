export * from "./load";
export * from "./nodes";
export * from "./walk";
