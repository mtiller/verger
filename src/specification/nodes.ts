export type FieldStruct = "scalar" | "optional" | "map" | "array" | "set";

export type BuiltinTypes = "string" | "number" | "boolean";

/** Covers singleton or union */
export interface BuiltinType {
  kind: "builtin";
  types: Array<BuiltinTypes>;
}

/** Represents a string enumeration */
export interface EnumType {
  kind: "enum";
  tags: string[];
}

/** Covers singleton or union */
export interface NodeType {
  kind: "node";
  types: string[];
}

/** The union of possible field types */
export type FieldType = BuiltinType | EnumType | NodeType;

export interface Field {
  type: FieldType;
  struct: FieldStruct;
}

/**
 * A subtype of `Field` where we know the field references
 * an AST Node (a child).
 */
export type NodeField = {
  type: NodeType;
  struct: FieldStruct;
};

/**
 * This is just a simple interface (that will not appear
 * in any union) for storing information about fields.
 **/
export interface ASTBaseType {
  name: string;
  extends: string[];
  fields: Map<string, Field>;
}

/**
 * This is used to represent a union type node
 */
export interface ASTUnionType {
  type: "union";
  name: string;
  subtypes: string[];
}

/** A predicate for identifying union nodes */
export function isUnion(a: ASTTree): a is ASTUnionType {
  return a.type === "union";
}

/**
 * This is used to represent leave nodes
 */
export interface ASTLeafType {
  type: "leaf";
  tag: string;
  rootUnion: ASTUnionType;
  parentUnion: ASTUnionType;
  name: string;
  extends: string[];
  fields: Map<string, Field>;
}

/** Predicate for identifying leaf nodes */
export function isLeaf(a: ASTTree): a is ASTLeafType {
  return a.type === "leaf";
}

/** This is our ASTTree type which is just a basic union of union types and leaf types. */
export type ASTTree = ASTUnionType | ASTLeafType;
