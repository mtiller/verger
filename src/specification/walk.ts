import {
  ASTBaseType,
  ASTLeafType,
  ASTSpec,
  ASTTree,
  ASTUnionType,
  FieldType,
} from "./nodes";

export function walkNode(
  n: string,
  rootUnion: string,
  content: any,
  types: ASTSpec
): ASTTree {
  if (content === null || content === undefined)
    throw new Error(`Missing content for ${name}`);
  if (Array.isArray(content)) {
    const ret: ASTLeafType = {
      type: "leaf",
      name: n,
      tag: n.toLowerCase(),
      rootUnion: rootUnion,
      extends: [],
      fields: new Map<string, FieldType>(),
    };

    walkFields(ret.name, content, types.bases, ret.extends, ret.fields);
    if (types.names.has(ret.name)) {
      throw new Error(`Multiple definitions for type ${ret.name}`);
    }
    if (validName(ret.name)) {
      types.names.add(ret.name);
      types.leaves.set(ret.name, ret);
    } else {
      throw new Error(`Invalid name ${ret.name} for AST node type`);
    }
    return ret;
  } else {
    const ret: ASTUnionType = {
      type: "union",
      name: n,
      subtypes: [],
    };
    for (const [subtype, contents] of Object.entries(content)) {
      if (!subtype.startsWith("^")) ret.subtypes.push(subtype);
      walkNode(subtype, rootUnion, contents, types);
    }
    if (types.names.has(ret.name)) {
      throw new Error(`Multiple definitions for type ${name}`);
    }
    if (validName(ret.name)) {
      types.names.add(ret.name);
      types.unions.set(ret.name, ret);
    } else {
      throw new Error(`Invalid name ${ret.name} for AST node type`);
    }
    return ret;
  }
}

export function walksBase(name: string, content: unknown, types: ASTSpec) {
  const base: ASTBaseType = {
    name: name,
    extends: [],
    fields: new Map(),
  };
  if (Array.isArray(content)) {
    walkFields(name, content, types.bases, base.extends, base.fields);
    types.names.add(name);
    types.bases.set(name, base);
  } else {
    throw new Error(
      `Expected contents of ${name} to be an array (of fields) but it wasn't`
    );
  }
}

function walkFields(
  className: string,
  content: any[],
  bases: Map<string, ASTBaseType>,
  supers: string[],
  fields: Map<string, FieldType>
) {
  for (const item of content) {
    const keys = Object.keys(item);
    if (keys.length !== 1) {
      throw new Error(
        `Unexpected field type specification: ${JSON.stringify(item)}`
      );
    } else {
      const fieldName = keys[0];
      const type = item[fieldName];
      if (typeof type === "string") {
        if (fieldName === "extends") {
          const base = bases.get(type);
          if (base === undefined) {
            throw new Error(
              `Type ${className} cannot extend from unknown (not yet defined) base type ${type}`
            );
          }
          supers.push(type);
        } else {
          if (type === "string" || type === "number" || type === "boolean") {
            fields.set(fieldName, { type: "scalar", name: type });
          } else {
            fields.set(fieldName, { type: "node", name: type });
          }
        }
      } else if (Array.isArray(type)) {
        if (type.every((x) => typeof x === "string")) {
          fields.set(fieldName, { type: "literals", tags: type });
        }
      } else {
        throw new Error(
          `Unexpected value for field ${fieldName}: ${JSON.stringify(type)}}`
        );
      }
    }
  }
}

const idRegExp = /^[$A-Z_][0-9A-Z_$]*$/i;
function validName(n: string): boolean {
  return idRegExp.test(n);
}
