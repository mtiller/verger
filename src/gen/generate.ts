import { ASTSpec } from "../specification";
import { leafOrBaseCode, unionCode } from "./templates";
import { lines } from "./text";

export function generate(spec: ASTSpec): string {
  const leaves = [...spec.leaves.values()];
  const bases = [...spec.bases.values()];
  const unions = [...spec.unions.values()];
  const leafDefs = leaves.map((a) => leafOrBaseCode(a, spec));
  const unionDefs = unions.map((a) => unionCode(a, spec));
  const stats = `//   Unions: ${unions.map((a) => a.name)}
  //   Bases: ${bases.map((a) => a.name)}
  //   Leaves: ${leaves.map((a) => a.name)}
  `;

  const preamble = `// DO NOT EDIT
  // This file was automatically generated
  
  ${stats}
  `;

  return lines(preamble, lines(...leafDefs), lines(...unionDefs));
}
