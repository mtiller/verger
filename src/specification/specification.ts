import { ASTBaseType, ASTLeafType, ASTUnionType } from "./nodes";

/** Represents options for the code generator */
export interface Options {
  tagName: string;
  optional: "json" | "expnull" | "purify";
  maps: "json" | "map";
}

/**
 * This is the complete specification of the abstract
 * syntax tree being described and then generated.
 */
export interface ASTSpec {
  names: Set<string>;
  unions: Map<string, ASTUnionType>;
  bases: Map<string, ASTBaseType>;
  leaves: Map<string, ASTLeafType>;
  options: Options;
}
