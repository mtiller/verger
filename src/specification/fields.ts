import { ASTSpec } from "./specification";
import { validName } from "./walk";

export type FieldStruct = "scalar" | "optional" | "map" | "array" | "set";

export type BuiltinTypes = "string" | "number" | "boolean";

/** Covers singleton or union */
export interface BuiltinType {
  kind: "builtin";
  types: Array<BuiltinTypes>;
}

export interface EnumType {
  kind: "enum";
  tags: string[];
}

/** Covers singleton or union */
export interface NodeType {
  kind: "node";
  types: string[];
}

export type NodeField = {
  type: NodeType;
  struct: FieldStruct;
};
export const isNodeField = (x: Field): x is NodeField => {
  return x.type.kind === "node";
};

export const isNodeFieldEntry = (
  x: [string, Field]
): x is [string, NodeField] => {
  return isNodeField(x[1]);
};

export interface Field {
  type: FieldType;
  struct: FieldStruct;
}
export type FieldType = BuiltinType | EnumType | NodeType;

function isBuiltin(x: string): x is BuiltinTypes {
  switch (x) {
    case "string":
    case "boolean":
    case "number":
      return true;
  }
  return false;
}

export function parseType(str: string, spec: ASTSpec): FieldType {
  const types = str.split("|").map((x) => x.trim());
  if (types.every((x) => validName(x))) {
    if (types.every(isBuiltin)) {
      return {
        kind: "builtin",
        types: types.filter(isBuiltin),
      };
    }
    if (types.every((x) => spec.names.has(x))) {
      return {
        kind: "node",
        types: types,
      };
    }
    throw new Error(`Unrecognized type name: '${str}'`);
  } else {
    throw new Error(`${str} is not a valid type name`);
  }
}
export function parseField(str: string, spec: ASTSpec): Field {
  if (str.endsWith("?")) {
    const type = str.slice(0, str.length - 1);
    return {
      struct: "optional",
      type: parseType(type, spec),
    };
  }
  if (str.endsWith("[]")) {
    const type = str.slice(0, str.length - 2);
    return {
      struct: "array",
      type: parseType(type, spec),
    };
  }
  return {
    struct: "scalar",
    type: parseType(str, spec),
  };
}
