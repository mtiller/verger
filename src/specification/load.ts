import { validName } from "../utils";
import { astUnionType, ASTUnionType, ASTSpec } from "./nodes";
import { walkNode } from "./walk";

/**
 * Load an AST specification by traversing data provided
 * (generally from YAML).
 */
export function loadSpec(ydata: any): ASTSpec {
  /** Create the spec we will return. */
  const spec: ASTSpec = {
    unions: new Map(),
    leaves: new Map(),
    externs: new Map(),
    options: {
      tagName: "tag",
      optional: "json",
      maps: "json",
      constructor: "inline",
    },
  };

  /** Parse any options first. */
  if (ydata.hasOwnProperty("options")) {
    loadOptions(ydata["options"], spec);
  }

  /** Create the nodes for bases (base class types) and nodes  */
  const nodes: any = ydata["nodes"];

  /** If no nodes are specified, this is an error */
  if (nodes === undefined) {
    throw new Error("Missing nodes field");
  }

  /** Walk nodes */
  for (const [name, content] of Object.entries(nodes)) {
    const rootUnion: ASTUnionType = astUnionType(name, []);

    walkNode(name, rootUnion, content, spec);
  }

  /** Return the resulting specification */
  return spec;
}

/**
 * A nice way to verify (in a type safe way) that a given
 * string belongs to a set of possible values.
 */
function assertOptions<K extends string>(
  name: string,
  possible: K[],
  actual: string
): K {
  if (possible.includes(actual as K)) return actual as K;
  if (typeof actual !== "string")
    throw new Error(`Expected string for option ${name}, got ${typeof actual}`);
  throw new Error(
    `For option '${name}' got '${actual}' but expected on of ${possible
      .map((x) => `'${x}'`)
      .join(", ")}`
  );
}

/**
 * Read in a collection of key-value pairs and use them to initial options.
 * @param options
 * @param spec
 */
export function loadOptions(options: any, spec: ASTSpec) {
  const map = new Map<string, string>();
  for (const [name, content] of Object.entries(options)) {
    if (typeof content !== "string")
      throw new Error(
        `Expected option ${name} to be a string, but it was '${typeof content}'`
      );
    map.set(name, content);
  }
  for (const [key, val] of map.entries()) {
    switch (key) {
      case "optional": {
        spec.options.optional = assertOptions(
          key,
          ["json", "expnull", "purify"],
          val
        );
        break;
      }
      case "tagName": {
        if (!validName(val)) {
          throw new Error(
            `The tagName option must be a valid Javascript identifier, but '${val}' is not.`
          );
        }
        spec.options.tagName = val;
        break;
      }
      case "maps": {
        spec.options.maps = assertOptions(key, ["json", "map"], val);
        break;
      }
      case "constructor": {
        spec.options.constructor = assertOptions(key, ["inline", "obj"], val);
      }
      default: {
        throw new Error(`Unknown option ${key}`);
      }
    }
  }
}
