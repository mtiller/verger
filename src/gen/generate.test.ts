import fs from "fs";
import path from "path";
import yaml from "yaml";
import { loadSpec } from "../specification";
import { generate } from "./generate";

describe("Test sample AST specifications", () => {
  ["expr", "repmin", "kitchen"].forEach((file) => {
    it(`should generate proper output for ${file}.yaml using json`, () => {
      const raw = fs.readFileSync(
        path.join(__dirname, "..", "..", "samples", `${file}.yaml`)
      );
      const spec = yaml.parse(raw.toString());
      spec.options.optional = "json";
      const types = loadSpec(spec);
      const output = generate(types);

      expect(output).toMatchSnapshot();
    });
    it(`should generate proper output for ${file}.yaml using purify`, () => {
      const raw = fs.readFileSync(
        path.join(__dirname, "..", "..", "samples", `${file}.yaml`)
      );
      const spec = yaml.parse(raw.toString());
      spec.options.optional = "purify";
      const types = loadSpec(spec);
      const output = generate(types);

      expect(output).toMatchSnapshot();
    });
  });
});
