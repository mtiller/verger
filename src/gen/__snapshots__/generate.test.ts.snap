// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test sample AST specifications should generate proper output for expr.yaml 1`] = `
"// DO NOT EDIT
// This file was automatically generated

export interface BinaryOp  {
    left: Expr;
    right: Expr;
}
export interface Literal  {
    tag: \\"literal\\";
    value: number;
}
export class Literal {
    static is = (x: Expr): x is Literal => { return x.tag===\\"literal\\" }
    static children = (x: Literal) => { return [] as const }
    static tag = \\"literal\\"
}
export interface ArithmeticOp extends BinaryOp {
    tag: \\"arithmeticop\\";
    op: \\"+\\" | \\"-\\" | \\"*\\" | \\"/\\";
}
export class ArithmeticOp {
    static is = (x: Expr): x is ArithmeticOp => { return x.tag===\\"arithmeticop\\" }
    static children = (x: ArithmeticOp) => { return [x.left, x.right] as const }
    static tag = \\"arithmeticop\\"
}
export interface UnaryOp  {
    tag: \\"unaryop\\";
    expr: Expr;
    op: \\"()\\" | \\"+\\" | \\"-\\";
}
export class UnaryOp {
    static is = (x: Expr): x is UnaryOp => { return x.tag===\\"unaryop\\" }
    static children = (x: UnaryOp) => { return [x.expr] as const }
    static tag = \\"unaryop\\"
}
export interface RelOp extends BinaryOp {
    tag: \\"relop\\";
    op: \\"+\\" | \\"-\\" | \\"*\\" | \\"/\\";
}
export class RelOp {
    static is = (x: Expr): x is RelOp => { return x.tag===\\"relop\\" }
    static children = (x: RelOp) => { return [x.left, x.right] as const }
    static tag = \\"relop\\"
}
export type Expr = Literal | ArithmeticOp | UnaryOp | RelOp;"
`;

exports[`Test sample AST specifications should generate proper output for expr.yaml using json 1`] = `
"// DO NOT EDIT
// This file was automatically generated

export interface BinaryOp  {
    left: Expr;
    right: Expr;
}
export interface Literal  {
    tag: \\"literal\\";
    value: number;
}
export class Literal {
    static is = (x: Expr): x is Literal => { return x.tag===\\"literal\\" }
    static children = (x: Literal) => { return [] as const }
    static tag = \\"literal\\"
}
export interface ArithmeticOp extends BinaryOp {
    tag: \\"arithmeticop\\";
    op: \\"+\\" | \\"-\\" | \\"*\\" | \\"/\\";
}
export class ArithmeticOp {
    static is = (x: Expr): x is ArithmeticOp => { return x.tag===\\"arithmeticop\\" }
    static children = (x: ArithmeticOp) => { return [x.left, x.right] as const }
    static tag = \\"arithmeticop\\"
}
export interface UnaryOp  {
    tag: \\"unaryop\\";
    expr: Expr;
    op: \\"()\\" | \\"+\\" | \\"-\\";
}
export class UnaryOp {
    static is = (x: Expr): x is UnaryOp => { return x.tag===\\"unaryop\\" }
    static children = (x: UnaryOp) => { return [x.expr] as const }
    static tag = \\"unaryop\\"
}
export interface RelOp extends BinaryOp {
    tag: \\"relop\\";
    op: \\"+\\" | \\"-\\" | \\"*\\" | \\"/\\";
}
export class RelOp {
    static is = (x: Expr): x is RelOp => { return x.tag===\\"relop\\" }
    static children = (x: RelOp) => { return [x.left, x.right] as const }
    static tag = \\"relop\\"
}
export type Expr = Literal | ArithmeticOp | UnaryOp | RelOp;"
`;

exports[`Test sample AST specifications should generate proper output for expr.yaml using purify 1`] = `
"// DO NOT EDIT
// This file was automatically generated

export interface BinaryOp  {
    left: Expr;
    right: Expr;
}
export interface Literal  {
    tag: \\"literal\\";
    value: number;
}
export class Literal {
    static is = (x: Expr): x is Literal => { return x.tag===\\"literal\\" }
    static children = (x: Literal) => { return [] as const }
    static tag = \\"literal\\"
}
export interface ArithmeticOp extends BinaryOp {
    tag: \\"arithmeticop\\";
    op: \\"+\\" | \\"-\\" | \\"*\\" | \\"/\\";
}
export class ArithmeticOp {
    static is = (x: Expr): x is ArithmeticOp => { return x.tag===\\"arithmeticop\\" }
    static children = (x: ArithmeticOp) => { return [x.left, x.right] as const }
    static tag = \\"arithmeticop\\"
}
export interface UnaryOp  {
    tag: \\"unaryop\\";
    expr: Expr;
    op: \\"()\\" | \\"+\\" | \\"-\\";
}
export class UnaryOp {
    static is = (x: Expr): x is UnaryOp => { return x.tag===\\"unaryop\\" }
    static children = (x: UnaryOp) => { return [x.expr] as const }
    static tag = \\"unaryop\\"
}
export interface RelOp extends BinaryOp {
    tag: \\"relop\\";
    op: \\"+\\" | \\"-\\" | \\"*\\" | \\"/\\";
}
export class RelOp {
    static is = (x: Expr): x is RelOp => { return x.tag===\\"relop\\" }
    static children = (x: RelOp) => { return [x.left, x.right] as const }
    static tag = \\"relop\\"
}
export type Expr = Literal | ArithmeticOp | UnaryOp | RelOp;"
`;

exports[`Test sample AST specifications should generate proper output for kitchen.yaml 1`] = `
"// DO NOT EDIT
// This file was automatically generated


export interface Function  {
    tag: \\"function\\";
    name: string;
    args: Argument[];
    body: Body;
}
export class Function {
    static is = (x: Node): x is Function => { return x.tag===\\"function\\" }
    static children = (x: Function) => { return [x.body, ...x.args] as const }
    static tag = \\"function\\"
}
export interface Declaration  {
    tag: \\"declaration\\";
    name: string;
    type: string;
    init: number;
}
export class Declaration {
    static is = (x: Node): x is Declaration => { return x.tag===\\"declaration\\" }
    static children = (x: Declaration) => { return [] as const }
    static tag = \\"declaration\\"
}
export interface Argument  {
    tag: \\"argument\\";
    name: string;
    type: string;
}
export class Argument {
    static is = (x: Node): x is Argument => { return x.tag===\\"argument\\" }
    static children = (x: Argument) => { return [] as const }
    static tag = \\"argument\\"
}
export interface Body  {
    tag: \\"body\\";
    statement: Statement;
}
export class Body {
    static is = (x: Node): x is Body => { return x.tag===\\"body\\" }
    static children = (x: Body) => { return [x.statement] as const }
    static tag = \\"body\\"
}
export interface AssignmentStatement  {
    tag: \\"assignmentstatement\\";
    name: string;
    init: number;
}
export class AssignmentStatement {
    static is = (x: Node): x is AssignmentStatement => { return x.tag===\\"assignmentstatement\\" }
    static children = (x: AssignmentStatement) => { return [] as const }
    static tag = \\"assignmentstatement\\"
}
export interface DeclarationStatement  {
    tag: \\"declarationstatement\\";
    name: string;
}
export class DeclarationStatement {
    static is = (x: Node): x is DeclarationStatement => { return x.tag===\\"declarationstatement\\" }
    static children = (x: DeclarationStatement) => { return [] as const }
    static tag = \\"declarationstatement\\"
}
export type Statement = AssignmentStatement | DeclarationStatement;
export type Node = Function | Declaration | Argument | Body | Statement;"
`;

exports[`Test sample AST specifications should generate proper output for kitchen.yaml using json 1`] = `
"// DO NOT EDIT
// This file was automatically generated


export interface Function  {
    tag: \\"function\\";
    name: string;
    args: Argument[];
    body: Body;
}
export class Function {
    static is = (x: Node): x is Function => { return x.tag===\\"function\\" }
    static children = (x: Function) => { return [x.body, ...x.args] as const }
    static tag = \\"function\\"
}
export interface Declaration  {
    tag: \\"declaration\\";
    name: string;
    type: string;
    init: number;
}
export class Declaration {
    static is = (x: Node): x is Declaration => { return x.tag===\\"declaration\\" }
    static children = (x: Declaration) => { return [] as const }
    static tag = \\"declaration\\"
}
export interface Argument  {
    tag: \\"argument\\";
    name: string;
    type: string;
}
export class Argument {
    static is = (x: Node): x is Argument => { return x.tag===\\"argument\\" }
    static children = (x: Argument) => { return [] as const }
    static tag = \\"argument\\"
}
export interface Body  {
    tag: \\"body\\";
    statement: Statement;
}
export class Body {
    static is = (x: Node): x is Body => { return x.tag===\\"body\\" }
    static children = (x: Body) => { return [x.statement] as const }
    static tag = \\"body\\"
}
export interface AssignmentStatement  {
    tag: \\"assignmentstatement\\";
    name: string;
    init: number;
}
export class AssignmentStatement {
    static is = (x: Node): x is AssignmentStatement => { return x.tag===\\"assignmentstatement\\" }
    static children = (x: AssignmentStatement) => { return [] as const }
    static tag = \\"assignmentstatement\\"
}
export interface DeclarationStatement  {
    tag: \\"declarationstatement\\";
    name: string;
}
export class DeclarationStatement {
    static is = (x: Node): x is DeclarationStatement => { return x.tag===\\"declarationstatement\\" }
    static children = (x: DeclarationStatement) => { return [] as const }
    static tag = \\"declarationstatement\\"
}
export type Statement = AssignmentStatement | DeclarationStatement;
export type Node = Function | Declaration | Argument | Body | Statement;"
`;

exports[`Test sample AST specifications should generate proper output for kitchen.yaml using purify 1`] = `
"// DO NOT EDIT
// This file was automatically generated


export interface Function  {
    tag: \\"function\\";
    name: string;
    args: Argument[];
    body: Body;
}
export class Function {
    static is = (x: Node): x is Function => { return x.tag===\\"function\\" }
    static children = (x: Function) => { return [x.body, ...x.args] as const }
    static tag = \\"function\\"
}
export interface Declaration  {
    tag: \\"declaration\\";
    name: string;
    type: string;
    init: number;
}
export class Declaration {
    static is = (x: Node): x is Declaration => { return x.tag===\\"declaration\\" }
    static children = (x: Declaration) => { return [] as const }
    static tag = \\"declaration\\"
}
export interface Argument  {
    tag: \\"argument\\";
    name: string;
    type: string;
}
export class Argument {
    static is = (x: Node): x is Argument => { return x.tag===\\"argument\\" }
    static children = (x: Argument) => { return [] as const }
    static tag = \\"argument\\"
}
export interface Body  {
    tag: \\"body\\";
    statement: Statement;
}
export class Body {
    static is = (x: Node): x is Body => { return x.tag===\\"body\\" }
    static children = (x: Body) => { return [x.statement] as const }
    static tag = \\"body\\"
}
export interface AssignmentStatement  {
    tag: \\"assignmentstatement\\";
    name: string;
    init: number;
}
export class AssignmentStatement {
    static is = (x: Node): x is AssignmentStatement => { return x.tag===\\"assignmentstatement\\" }
    static children = (x: AssignmentStatement) => { return [] as const }
    static tag = \\"assignmentstatement\\"
}
export interface DeclarationStatement  {
    tag: \\"declarationstatement\\";
    name: string;
}
export class DeclarationStatement {
    static is = (x: Node): x is DeclarationStatement => { return x.tag===\\"declarationstatement\\" }
    static children = (x: DeclarationStatement) => { return [] as const }
    static tag = \\"declarationstatement\\"
}
export type Statement = AssignmentStatement | DeclarationStatement;
export type Node = Function | Declaration | Argument | Body | Statement;"
`;

exports[`Test sample AST specifications should generate proper output for repmin.yaml 1`] = `
"// DO NOT EDIT
// This file was automatically generated


export interface Fork  {
    tag: \\"fork\\";
    left: Tree;
    right: Tree;
}
export class Fork {
    static is = (x: Tree): x is Fork => { return x.tag===\\"fork\\" }
    static children = (x: Fork) => { return [x.left, x.right] as const }
    static tag = \\"fork\\"
}
export interface Leaf  {
    tag: \\"leaf\\";
    value: number;
}
export class Leaf {
    static is = (x: Tree): x is Leaf => { return x.tag===\\"leaf\\" }
    static children = (x: Leaf) => { return [] as const }
    static tag = \\"leaf\\"
}
export type Tree = Fork | Leaf;"
`;

exports[`Test sample AST specifications should generate proper output for repmin.yaml using json 1`] = `
"// DO NOT EDIT
// This file was automatically generated


export interface Fork  {
    tag: \\"fork\\";
    left: Tree;
    right: Tree;
}
export class Fork {
    static is = (x: Tree): x is Fork => { return x.tag===\\"fork\\" }
    static children = (x: Fork) => { return [x.left, x.right] as const }
    static tag = \\"fork\\"
}
export interface Leaf  {
    tag: \\"leaf\\";
    value: number;
}
export class Leaf {
    static is = (x: Tree): x is Leaf => { return x.tag===\\"leaf\\" }
    static children = (x: Leaf) => { return [] as const }
    static tag = \\"leaf\\"
}
export type Tree = Fork | Leaf;"
`;

exports[`Test sample AST specifications should generate proper output for repmin.yaml using purify 1`] = `
"// DO NOT EDIT
// This file was automatically generated


export interface Fork  {
    tag: \\"fork\\";
    left: Tree;
    right: Tree;
}
export class Fork {
    static is = (x: Tree): x is Fork => { return x.tag===\\"fork\\" }
    static children = (x: Fork) => { return [x.left, x.right] as const }
    static tag = \\"fork\\"
}
export interface Leaf  {
    tag: \\"leaf\\";
    value: number;
}
export class Leaf {
    static is = (x: Tree): x is Leaf => { return x.tag===\\"leaf\\" }
    static children = (x: Leaf) => { return [] as const }
    static tag = \\"leaf\\"
}
export type Tree = Fork | Leaf;"
`;
